dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/AbstractEvent.java

```diff
@@ -160,21 +144,55 @@ public final List<Event> getPredecessors() {
         return events;
     }
 
+    /*
+        Detaches this event from the control-flow graph.
+        This does not properly delete the event, and it may be reinserted elsewhere.
+        TODO: We need to special-case handle detaching the entry/exit event of a thread.
+     */
+    @Override
+    public void detach() {
+        Preconditions.checkState(thread == null || (this != thread.getEntry() && this != thread.getExit()),
```

Comment:
It checks an implication `e.hasThread => e is neither entry nor exit event of that thread`. Since we cannot formulate implications directly, I transform this into `!e.hasThread OR e is neither entry nor exit event of that thread`.
I kinda want to get rid of this check and allow entry/exit to get detached, but then `thread.entry/exit` have to get updated and the former is currently not accessible (there is no `thread.setEntry` ).
So it was easier for now to make sure that entry and exit are not detachable.

Reply:
the question is: do we need the precondition? can this method be called for an event that does not belong to any thread?

Reply:
The code can be called on thread-less events, for example those that were returned by `EventFactory.newXYZ` or those that were just detached (you can call `detach` multiple times in a row).
Removing the `thread == null` check will just change the current `IllegalStateException` into a `NullPointerException`, so you don't get anything.
The checks for `entry/exit` are important because detaching those will result in inconsistent state that breaks down-the-line. That's not debuggable.